### PR TITLE
Check that ipfs and ipns hashes correspond

### PR DESCRIPTION
**Pull request by meditans**

### CODE REVIEW

File: src/libstore/ipfs-binary-cache-store.cc [hunk 1]
```diff
@@ -92,6 +86,17 @@ class IPFSBinaryCacheStore : public BinaryCacheStore
...
+    // Given a ipns path, checks if it corresponds to a DNSLink path.
+    bool isDNSLinkPath(std::string path) {
+        if (path.find("/ipns/") != 0) {
+            auto subpath = std::string(path, 6);
+            if (subpath.find(".") == std::string::npos)
+                return false;
+            return true;
+        }
+        throw Error("The provided path is not a ipns path");
+    }
```
Comment (Ericson2314) [inline src/libstore/ipfs-binary-cache-store.cc:98]:
```suggestion
    bool isDNSLinkPath(std::string path) {
        if (path.find("/ipns/") == 0)
            throw Error("The provided path is not a ipns path");
        auto subpath = std::string(path, 6);
        return subpath.find(".") != std::string::npos;
    }
```

File: src/libstore/ipfs-binary-cache-store.cc [hunk 2]
```diff
@@ -111,14 +116,41 @@ class IPFSBinaryCacheStore : public BinaryCacheStore
...
         auto ipnsPath = *optIpnsPath;
 
-        auto state(_state.lock());
+        if (isDNSLinkPath(ipnsPath)) {
+            throw Error("The provided ipns path is a DNSLink, and syncing those is not supported.\n  ipns path: %s", ipnsPath);
+        }
+
+        auto resolvedIpfsPath = resolveIPNSName(ipnsPath, false);
+        if (resolvedIpfsPath != initialIpfsPath) {
+            throw Error("The ipns hash %s doesn't correspond to the initial ipfs hash;\n  wanted: %s\n  got %s",
```
Comment (Ericson2314) [inline src/libstore/ipfs-binary-cache-store.cc:148]: Mention
```suggestion
            throw Error("The IPNS hash or DNS link %s doesn't now resolves to something different than when Nix was started;\n  wanted: %s\n  got %s\nPerhaps thing else updated it in the meantime?",
```

File: src/libstore/ipfs-binary-cache-store.cc [hunk 3]
```diff
@@ -111,14 +116,41 @@ class IPFSBinaryCacheStore : public BinaryCacheStore
...
+            throw Error("We don't have an ipns path and the current ipfs address doesn't match the initial one.\n  current: %s\n  initial: %s",
+                state->ipfsPath, initialIpfsPath);
+        }
+
         auto ipnsPath = *optIpnsPath;
 
-        auto state(_state.lock());
+        if (isDNSLinkPath(ipnsPath)) {
+            throw Error("The provided ipns path is a DNSLink, and syncing those is not supported.\n  ipns path: %s", ipnsPath);
+        }
```
Comment (Ericson2314) [inline src/libstore/ipfs-binary-cache-store.cc:144]: This should go below the next new bit, and like the above mention the two values, and also instruct the user to update their DNS by hand.

File: src/libstore/ipfs-binary-cache-store.cc (hunk 2 above)
Comment (Ericson2314) [reply src/libstore/ipfs-binary-cache-store.cc:148]: Also use "initial" and "current" here.

File: src/libstore/ipfs-binary-cache-store.cc (hunk 3 above)
Comment (meditans) [reply src/libstore/ipfs-binary-cache-store.cc]: Done, thanks

**Pull request merged**